File: solutions/cpp/Search_a_pair_sorted_array.cpp
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <functional>
#include <iostream>
#include <random>
#include <utility>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::greater_equal;
using std::less;
using std::pair;
using std::random_device;
using std::uniform_int_distribution;
using std::vector;

template <typename Comp>
pair<int, int> FindPairUsingComp(const vector<int>& A, int k, Comp comp);
pair<int, int> FindPositiveNegativePair(const vector<int>& A, int k);

// @include
pair<int, int> FindPairSumK(const vector<int>& A, int k) {
  pair<int, int> ret = FindPositiveNegativePair(A, k);
  if (ret.first == -1 && ret.second == -1) {
    return k >= 0 ? FindPairUsingComp(A, k, less<int>())
                  : FindPairUsingComp(A, k, greater_equal<int>());
  }
  return ret;
}

template <typename Comp>
pair<int, int> FindPairUsingComp(const vector<int>& A, int k, Comp comp) {
  pair<int, int> ret(0, A.size() - 1);
  while (ret.first < ret.second && comp(A[ret.first], 0)) {
    ++ret.first;
  }
  while (ret.first < ret.second && comp(A[ret.second], 0)) {
    --ret.second;
  }

  while (ret.first < ret.second) {
    if (A[ret.first] + A[ret.second] == k) {
      return ret;
    } else if (comp(A[ret.first] + A[ret.second], k)) {
      do {
        ++ret.first;
      } while (ret.first < ret.second && comp(A[ret.first], 0));
    } else {
      do {
        --ret.second;
      } while (ret.first < ret.second && comp(A[ret.second], 0));
    }
  }
  return {-1, -1};  // no answer.
}

pair<int, int> FindPositiveNegativePair(const vector<int>& A, int k) {
  // ret.first for positive, and ret.second for negative.
  pair<int, int> ret(A.size() - 1, A.size() - 1);
  // Find the last positive or zero.
  while (ret.first >= 0 && A[ret.first] < 0) {
    --ret.first;
  }

  // Find the last negative.
  while (ret.second >= 0 && A[ret.second] >= 0) {
    --ret.second;
  }

  while (ret.first >= 0 && ret.second >= 0) {
    if (A[ret.first] + A[ret.second] == k) {
      return ret;
    } else if (A[ret.first] + A[ret.second] > k) {
      do {
        --ret.first;
      } while (ret.first >= 0 && A[ret.first] < 0);
    } else {  // A[ret.first] + A[ret.second] < k.
      do {
        --ret.second;
      } while (ret.second >= 0 && A[ret.second] >= 0);
    }
  }
  return {-1, -1};  // no answer.
}
// @exclude

int main(int argc, char* argv[]) {
  default_random_engine gen((random_device())());
  for (int times = 0; times < 10000; ++times) {
    int n;
    if (argc >= 2) {
      n = atoi(argv[1]);
    } else {
      uniform_int_distribution<int> dis(1, 10000);
      n = dis(gen);
    }
    vector<int> A;
    uniform_int_distribution<int> pos_or_neg(0, 1);
    uniform_int_distribution<int> dis(0, 10000);
    for (size_t i = 0; i < n; ++i) {
      A.emplace_back(((pos_or_neg(gen)) ? 1 : -1) * (dis(gen)));
    }
    sort(A.begin(), A.end(), [](int x, int y) { return abs(x) < abs(y); });
    int k = ((pos_or_neg(gen)) ? 1 : -1) * (dis(gen));
    /*
    for (const int& a : A) {
      cout << a << " ";
    }
    cout << endl << "k = " << k << endl;
    */
    pair<int, int> ans = FindPairSumK(A, k);
    if (ans.first != -1 && ans.second != -1) {
      assert(A[ans.first] + A[ans.second] == k);
      cout << A[ans.first] << "+" << A[ans.second] << "=" << k << endl;
    } else {
      sort(A.begin(), A.end());
      int l = 0, r = A.size() - 1;
      bool found = false;
      while (l < r) {
        if (A[l] + A[r] == k) {
          cout << A[l] << "+" << A[r] << "=" << k << endl;
          found = true;
          break;
        } else if (A[l] + A[r] < k) {
          ++l;
        } else {
          --r;
        }
      }
      cout << "no answer" << endl;
      assert(!found);
    }
  }
  return 0;
}

File: static/solutions/cpp/java/src/main/java/com/epi/SearchAPairSortedArray.java
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

package com.epi;

import java.util.Arrays;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Random;

public class SearchAPairSortedArray {
  // @include
  private static class IndexPair {
    public Integer index1;
    public Integer index2;

    public IndexPair(Integer index1, Integer index2) {
      this.index1 = index1;
      this.index2 = index2;
    }
  }

  private static interface BooleanCompare {
    public boolean compare(Integer index1, Integer index2);
  }

  private static class CompareLess implements BooleanCompare {
    // clang-format off
    @Override
    public boolean compare(Integer o1, Integer o2) { return o1 < o2; }
    // clang-format on
    public static final CompareLess LESS = new CompareLess();
  }

  private static class CompareGreaterEqual implements BooleanCompare {
    // clang-format off
    @Override
    public boolean compare(Integer o1, Integer o2) { return o1 >= o2; }
    // clang-format on
    public static final CompareGreaterEqual GREATER_OR_EQUAL
        = new CompareGreaterEqual();
  }

  public static IndexPair findPairSumK(List<Integer> A, int k) {
    IndexPair result = findPosNegPair(A, k);
    if (result.index1 == -1 && result.index2 == -1) {
      return k >= 0
          ? findPairUsingComp(A, k, CompareLess.LESS)
          : findPairUsingComp(A, k, CompareGreaterEqual.GREATER_OR_EQUAL);
    }
    return result;
  }

  private static IndexPair findPairUsingComp(List<Integer> A, int k,
                                             BooleanCompare comp) {
    IndexPair result = new IndexPair(0, A.size() - 1);
    while (result.index1 < result.index2
           && comp.compare(A.get(result.index1), 0)) {
      result.index1 = result.index1 + 1;
    }
    while (result.index1 < result.index2
           && comp.compare(A.get(result.index2), 0)) {
      result.index2 = result.index2 - 1;
    }

    while (result.index1 < result.index2) {
      if (A.get(result.index1) + A.get(result.index2) == k) {
        return result;
      } else if (comp.compare(A.get(result.index1) + A.get(result.index2), k)) {
        do {
          result.index1 = result.index1 + 1;
        } while (result.index1 < result.index2
                 && comp.compare(A.get(result.index1), 0));
      } else {
        do {
          result.index2 = result.index2 - 1;
        } while (result.index1 < result.index2
                 && comp.compare(A.get(result.index2), 0));
      }
    }
    return new IndexPair(-1, -1); // No answer.
  }

  private static IndexPair findPosNegPair(List<Integer> A, int k) {
    // result.first for positive, and result.second for negative.
    IndexPair result = new IndexPair(A.size() - 1, A.size() - 1);
    // Find the last positive or zero.
    while (result.index1 >= 0 && A.get(result.index1) < 0) {
      result.index1 = result.index1 - 1;
    }

    // Find the last negative.
    while (result.index2 >= 0 && A.get(result.index2) >= 0) {
      result.index2 = result.index2 - 1;
    }

    while (result.index1 >= 0 && result.index2 >= 0) {
      if (A.get(result.index1) + A.get(result.index2) == k) {
        return result;
      } else if (A.get(result.index1) + A.get(result.index2) > k) {
        do {
          result.index1 = result.index1 - 1;
        } while (result.index1 >= 0 && A.get(result.index1) < 0);
      } else { // A.get(result.first) + A.get(result.second) < k.
        do {
          result.index2 = result.index2 - 1;
        } while (result.index2 >= 0 && A.get(result.index2) >= 0);
      }
    }
    return new IndexPair(-1, -1); // No answer.
  }
  // @exclude

  private static void simpleTest() {
    IndexPair ans = findPairSumK(Arrays.asList(0, 0, -1, 2, -3, -3), 2);
    assert(ans.index1 != -1);
  }

  public static void main(String[] args) {
    simpleTest();
    Random rand = new Random();
    for (int times = 0; times < 10000; ++times) {
      int n;
      if (args.length >= 1) {
        n = Integer.parseInt(args[0]);
      } else {
        n = rand.nextInt(10000) + 1;
      }
      List<Integer> A = new ArrayList<>();
      for (int i = 0; i < n; ++i) {
        A.add(rand.nextInt(19999) - 9999);
      }
      Collections.sort(A, new Comparator<Integer>() {
        @Override
        public int compare(Integer o1, Integer o2) {
          return Integer.compare(Math.abs(o1), Math.abs(o2));
        }
      });
      int k = rand.nextInt(19999) - 9999;
      IndexPair ans = findPairSumK(A, k);
      if (ans.index1 != -1 && ans.index2 != -1) {
        assert(A.get(ans.index1) + A.get(ans.index2) == k);
        System.out.println(A.get(ans.index1) + "+" + A.get(ans.index2) + "="
                           + k);
      } else {
        Collections.sort(A);
        int l = 0, r = A.size() - 1;
        boolean found = false;
        while (l < r) {
          if (A.get(l) + A.get(r) == k) {
            System.out.println(A.get(l) + "+" + A.get(r) + "=" + k);
            Collections.sort(A, new Comparator<Integer>() {
              @Override
              public int compare(Integer o1, Integer o2) {
                return Integer.compare(Math.abs(o1), Math.abs(o2));
              }
            });
            ans = findPairSumK(A, k);
            found = true;
            break;
          } else if (A.get(l) + A.get(r) < k) {
            ++l;
          } else {
            --r;
          }
        }
        System.out.println("no answer");
        assert(!found);
      }
    }
  }
}
